python/ray/data/datasource/file_based_datasource.py

```diff
@@ -467,6 +507,55 @@ def read_files(
         return read_tasks
 
 
+def _add_partitions(
+    data: Union["pyarrow.Table", "pd.DataFrame"], partitions: Dict[str, Any]
+) -> Union["pyarrow.Table", "pd.DataFrame"]:
+    import pandas as pd
+    import pyarrow as pa
+
+    assert isinstance(data, (pa.Table, pd.DataFrame))
+    if isinstance(data, pa.Table):
+        return _add_partitions_to_table(data, partitions)
+    if isinstance(data, pd.DataFrame):
+        return _add_partitions_to_dataframe(data, partitions)
+
+
+def _add_partitions_to_table(
+    table: "pyarrow.Table", partitions: Dict[str, Any]
+) -> "pyarrow.Table":
+    for field in table.column_names:
+        if field in partitions:
+            raise RuntimeError(
+                f"{field} is a partition key, but it's also the name of a column in the"
+                "read dataset."
+            )
```

Comment:
Hmm I'm not sure if we'd want to error here actually, there's a semi-common pattern that we've encountered with users where their partition columns exist both in the file paths AND in the actual data, and they expect this to work without issue.

Instead, I think that it would be better to validate that the partition value and the in-data values are consistent (raise an error if not), and then leave the existing in-data column as-is.